### ISSUE TITLE
Should use a canvas rather than divs?

### ISSUE THREAD

**Issue by CurtisFulton** [camerongillette/factorio_blueprint_editor#25]
I'm not sure about performance, but would it be a better idea to use a canvas to render the blueprints? It would make it easier to zoom in and out.

I'm just not sure how good performance is with a canvas (Haven't done much with them). Would this be a better option?

**Comment by demipixel** [comment]
WebGL is gonna have better performance than canvas.

**Comment by CurtisFulton** [comment]
WebGL still uses a canvas though doesn't it? My OpenGL is a bit rusty, but I can try to make something for this if it's worth it.

**Comment by demipixel** [comment]
To put it better, you can make more optimizations in webGL and have more power, whereas canvas leaves most optimizations up to the browser. For a basic comparison:

https://www.scirra.com/labs/perftest-2d/
https://www.scirra.com/labs/perftest-webgl/

But this is only the tip of the iceberg, considering how much faster GPUs run vs CPUs (and how webGL can interact more directly with the GPU).

EDIT: Beware of mobile support, will have to look into this...

**Comment by CurtisFulton** [comment]
Funnily enough my browser runs the non webgl at about 2-3x the performance.  Would this be something we can implement without WebGL, then add webGL optimizations at a later date if needed?

**Comment by demipixel** [comment]
Not very easily I don't think. Interesting that you get that performance, you should google the differences in performance and what other people say.

But I think there's a reason things like Google maps use webGL over canvas.

**Comment by CurtisFulton** [comment]
That's true. I think WebGL would have better performance because you would be able to use UV mapping to draw the images. 

At the same time though, just using a standard canvas shouldn't be very bad performance (Unless we do panning/zooming) because you wouldn't have to constantly re-draw the whole blueprint. Once an image is drawn, you just don't clear it. The only part that has to be re-drawn is where the mouse is for the ghost image (And 1 re-draw to draw the image in the grid when it is placed).

**Comment by demipixel** [comment]
Depends if you want to [eventually move to panning/zooming and realistic graphics](https://cdn.discordapp.com/attachments/314607309508575236/382467053170196481/blueprint.png).

**Comment by CurtisFulton** [comment]
Doing realistic graphics is nice, but it introduces a few additional challenges like draw order (powerpoles for example). 

I think I'll start on a canvas implementation and see how I go. It might end up being better to work on #16 first before worrying about changing the rendering system.

**Comment by demipixel** [comment]
This is a little bit off-topic from this discussion, but I've taken my own stab at realistic rendering to make an editor that feels like Factorio. Draw order is the least of your worries.

**Comment by ryepup** [comment]
There's also https://github.com/trakos/fbpviewer-js hanging out there, might be able to integrate with that
